[jit] improve error message on inferred type

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21058 [jit] improve error message on inferred type**

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

This PR moves the type inference logic from the frontend to
compiler.cpp. So now the `Param` tree view's type is optional, and we
track whether or not we inferred the type of a schema argument to be
Tensor. We use that info to format a proper error message.

Also this dedupes some of the error messages reporting since we were
replicating similar strings in several places.

gh-metadata: pytorch pytorch 21058 gh/suo/42/head

Differential Revision: [D15547218](https://our.internmc.facebook.com/intern/diff/D15547218)

suo committed:
[jit] improve error message on inferred type

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

suo committed:
Update on "[jit] improve error message on inferred type"

[jit] improve error message on inferred type

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

This PR moves the type inference logic from the frontend to
compiler.cpp. So now the `Param` tree view's type is optional, and we
track whether or not we inferred the type of a schema argument to be
Tensor. We use that info to format a proper error message.

Also this dedupes some of the error messages reporting since we were
replicating similar strings in several places.

gh-metadata: pytorch pytorch 21058 gh/suo/42/head

suo committed:
Update on "[jit] improve error message on inferred type"

[jit] improve error message on inferred type

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

This PR moves the type inference logic from the frontend to
compiler.cpp. So now the `Param` tree view's type is optional, and we
track whether or not we inferred the type of a schema argument to be
Tensor. We use that info to format a proper error message.

Also this dedupes some of the error messages reporting since we were
replicating similar strings in several places.

gh-metadata: pytorch pytorch 21058 gh/suo/42/head

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -49,6 +51,28 @@ struct Argument {
+    }
+    return c10::str(
+        "expected a value of type '",
+        type()->str(),
```
`python_str()`

zdevito commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -42,20 +42,39 @@ inline std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema)
+    TORCH_CHECK(
+        false,
+        formatTypeMismatchMsg(
+            argument, attemptToRecoverType(value)->str(), pos));
```
python_str() (a lot of our existing code has the problem, but the user should only ever see python_str())

zdevito reviewed (approved):
Looks good!

suo committed:
fix test/comments on "[jit] improve error message on inferred type"

[jit] improve error message on inferred type

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

This PR moves the type inference logic from the frontend to
compiler.cpp. So now the `Param` tree view's type is optional, and we
track whether or not we inferred the type of a schema argument to be
Tensor. We use that info to format a proper error message.

Also this dedupes some of the error messages reporting since we were
replicating similar strings in several places.

gh-metadata: pytorch pytorch 21058 gh/suo/42/head

suo committed:
Update on "[jit] improve error message on inferred type"

[jit] improve error message on inferred type

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

This PR moves the type inference logic from the frontend to
compiler.cpp. So now the `Param` tree view's type is optional, and we
track whether or not we inferred the type of a schema argument to be
Tensor. We use that info to format a proper error message.

Also this dedupes some of the error messages reporting since we were
replicating similar strings in several places.

gh-metadata: pytorch pytorch 21058 gh/suo/42/head

suo committed:
Update on "[jit] improve error message on inferred type"

[jit] improve error message on inferred type

Previously we would give confusing error messages when we inferred an
unannotated type to be Tensor. This adds a hint to the compiler
diagnostic so that users might be less confused.

This PR moves the type inference logic from the frontend to
compiler.cpp. So now the `Param` tree view's type is optional, and we
track whether or not we inferred the type of a schema argument to be
Tensor. We use that info to format a proper error message.

Also this dedupes some of the error messages reporting since we were
replicating similar strings in several places.

gh-metadata: pytorch pytorch 21058 gh/suo/42/head
